metadata-io/src/test/java/com/linkedin/metadata/graph/elastic/ElasticSearchGraphServiceTest.java

```diff
@@ -227,4 +230,142 @@ public void testConcurrentRemoveNodes() {
   }
+
+  @Test
+  public void testTimestampLineage() throws Exception{
+    // Populate one upstream and two downstream edges at initialTime
+    Long initialTime = 1000L;
+
+    List<Edge> edges = Arrays.asList(
+        new Edge(datasetTwoUrn, datasetOneUrn, downstreamOf, initialTime, null, initialTime, null, null),
+        new Edge(datasetThreeUrn, datasetTwoUrn, downstreamOf, initialTime, null, initialTime, null, null),
+        new Edge(datasetFourUrn, datasetTwoUrn, downstreamOf, initialTime, null, initialTime, null, null)
+    );
+
+    edges.forEach(getGraphService()::addEdge);
+    syncAfterWrite();
+
+    // Without timestamps
+    EntityLineageResult upstreamResult = getUpstreamLineage(datasetTwoUrn, null, null);
+    EntityLineageResult downstreamResult = getDownstreamLineage(datasetTwoUrn, null, null);
+    Assert.assertEquals(new Integer(1), upstreamResult.getTotal());
+    Assert.assertEquals(new Integer(2), downstreamResult.getTotal());
+
+    // Timestamp before
+    upstreamResult = getUpstreamLineage(datasetTwoUrn,
+        0L,
+        initialTime - 10);
+    downstreamResult = getDownstreamLineage(datasetTwoUrn,
+        0L,
+        initialTime - 10);
+    Assert.assertEquals(new Integer(0), upstreamResult.getTotal());
+    Assert.assertEquals(new Integer(0), downstreamResult.getTotal());
+
+    // Timestamp after
+    upstreamResult = getUpstreamLineage(datasetTwoUrn,
+        initialTime + 10,
+        initialTime + 100);
+    downstreamResult = getDownstreamLineage(datasetTwoUrn,
+        initialTime + 10,
+        initialTime + 100);
+    Assert.assertEquals(new Integer(0), upstreamResult.getTotal());
+    Assert.assertEquals(new Integer(0), downstreamResult.getTotal());
+
+    // Timestamp included
+    upstreamResult = getUpstreamLineage(datasetTwoUrn,
+        initialTime - 10,
+        initialTime + 10);
+    downstreamResult = getDownstreamLineage(datasetTwoUrn,
+        initialTime - 10,
+        initialTime + 10);
+    Assert.assertEquals(new Integer(1), upstreamResult.getTotal());
+    Assert.assertEquals(new Integer(2), downstreamResult.getTotal());
+
+    // Update only one of the downstream edges
+    Long updatedTime = 2000L;
+    edges = Arrays.asList(
+        new Edge(datasetTwoUrn, datasetOneUrn, downstreamOf, initialTime, null, updatedTime, null, null),
+        new Edge(datasetThreeUrn, datasetTwoUrn, downstreamOf, initialTime, null, updatedTime, null, null)
+    );
+
+    edges.forEach(getGraphService()::addEdge);
+    syncAfterWrite();
+
+    // Without timestamps
+    upstreamResult = getUpstreamLineage(datasetTwoUrn,
+        null,
+        null);
+    downstreamResult = getDownstreamLineage(datasetTwoUrn,
+        null,
+        null);
+    Assert.assertEquals(new Integer(1), upstreamResult.getTotal());
+    Assert.assertEquals(new Integer(2), downstreamResult.getTotal());
+
+    // Window includes initial time and updated time
+    upstreamResult = getUpstreamLineage(datasetTwoUrn,
```

Comment:
Nice! Thanks for capturing all of these cases.

Are we able to add one where both createdAt and updatedAt timestamps are a) null or b) 0? We should treat either case as representing a "non-existant edge" (and disqualify it).

This is where we've had issues in the past! cc @aditya-radhakrishnan

Reply:
Yes! So just to clarify, these should not be returned for any searches no matter the timestamp, nor for searches with null timestamps?

Reply:
Hmm actually @jjoyce0510 I don't think this is right. If both createdAt or updatedAt is 0 or null (falsey), we should return the edge.

Reply:
Sorry - meant what aditya said. If BOTH have null or 0 value, we always want to include that edge in result set

Reply:
Great I will have a test for this!